cirq-core/cirq/ops/conditional_operation.py

```diff
@@ -0,0 +1,153 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import (
+    AbstractSet,
+    Any,
+    cast,
+    Dict,
+    FrozenSet,
+    Optional,
+    Sequence,
+    TYPE_CHECKING,
+    Tuple,
+    Union,
+)
+
+from cirq import protocols, value
+from cirq.ops import raw_types
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality
+class ConditionalOperation(raw_types.Operation):
+    """Augments existing operations to be conditionally executed."""
+
+    def __init__(
+        self,
+        sub_operation: 'cirq.Operation',
+        keys: Union[str, 'cirq.MeasurementKey', Sequence[Union[str, 'cirq.MeasurementKey']]],
+    ):
+        keys = [keys] if isinstance(keys, (str, value.MeasurementKey)) else keys
+        keys = tuple(value.MeasurementKey(k) if isinstance(k, str) else k for k in keys)
+        self._control_keys = cast(Tuple['cirq.MeasurementKey', ...], keys)
+        self._sub_operation = sub_operation
+
+    @property
+    def conditions(self) -> FrozenSet['cirq.MeasurementKey']:
```

Comment:
Consider returning a set of a new custom `Condition` interface here. In the future, we'll have more complicated conditions. Make this fairly minimal. For now, there should be `Condition.keys()`, and possibly that's all.

Reply:
Okay, so https://github.com/daxfohl/Cirq/compare/scope3...daxfohl:sympy?expand=1 shows roughly what the `Condition` will look like with the sympy conditions. A few options here.

1. We could integrate all that into this PR. That's a fairly large increase in scope (we don't even have solid requirements on that yet) and I'd probably not want to.
2. We could just create a skeleton `Condition` and use that here. I'm a bit reluctant on this because even a skeleton Condition is a fair amount of extra work (serialization, testing, etc) and it may not even match the final sympy code.
3. We could leave this as-is and just make it a breaking change when/if sympy or other condition expressions get integrated.
4. We could rename this to `control_keys` and then deprecate it (or not) once we have a proper Condition class for a proper `conditions` property.

I somewhat lean toward (4) right now. One caveat of this is that `op.control_keys()` will be different than `protocols.control_keys(op)`, once the scoping PR gets merged. But we've talked about this before and it seems we're okay with that. The other caveat is that I do still think we should keep `with_conditions` and `unconditionally` as-is, and the name mismatch could be confusing. But that would only be short-term until we get a proper Condition class and method, so I think it's okay.

Reply:
Even better, I decided just to delete it. We do have a `control_keys` protocol, so we can use that instead to get control keys. Then we can add the `conditions` method once it makes sense to do so.